src/PolarSSL++/BlockingSslClientSocket.cpp

```diff
@@ -239,9 +243,16 @@ void cBlockingSslClientSocket::Disconnect(void)
 	}
 	
 	m_Ssl.NotifyClose();
-	m_Socket->Close();
-	m_Socket.reset();
 	m_IsConnected = false;
+
+	// Grab a copy of the socket so that we know it doesn't change under our hands:
+	auto socket = m_Socket;
+	if (socket != nullptr)
+	{
+		socket->Close();
+	}
+
+	m_Socket.reset();
```

Comment:
With this arrangement, is the `!m_IsConnected` check needed any more, considering that the lack of locking means that the variable and the state it's associated with could change however it wants after that line?

Also, would it be valid to move the `::reset` call into the `if`'s codeblock for performances and stuff?

Also also, `socket` -> `Socket`? :P